internal/utils/graph_utils.go

```diff
@@ -0,0 +1,41 @@
+package utils
+
+import (
+	"context"
+)
+
+type (
+	EngineUtils interface {
+		CheckVisited(ctx context.Context, current string) (context.Context, bool)
+	}
+	EngineUtilsProvider struct {
+		EngineUtils
+	}
```

Comment:
I don't really see why we would need to have a `EngineUtilsProvider`. It does not hold state and has no dependencies. Let's just use the function as is without a function receiver.